src/app/services/account/account.service.ts

```diff
@@ -1,81 +1,358 @@
+            { email, password }
         ).pipe(
-            first(), //emitt 1 item and close the Observable 
             tap(user => {
                 localStorage.setItem(
                     'currentUser',
                     JSON.stringify(user)
                 );
-                this.updateLocalStorageSubject(); 
-                return user; 
-            })
-        ); 
+                this.updateLocalStorageSubject();
+                return user;
+            }),
+            catchError(error => this.handleError(
+                error,
+                'User name or password is incorrect!'
+            ))
+        );
     }
 
-    logOut():void {
+    logOut(): void {
 
-  
         localStorage.removeItem('currentUser');
         //this.currentUserSubject.next("default");
         this.localStorageUserSubject.next('default');
         this.router.navigate(['/login']);
     }
 
-    signUp(userData:User) {
+    signUp(userData: User) {
         return this.http.post<User>(
             'api/register',
             userData
         ).
             pipe(
                 tap(item => console.log(item)),
                 catchError(error => {
-                    throw 'Email already exists in database!';
+                    return this.handleError(
+                        error,
+                        'Email already exists in database!'
+                    );
                 })
-            ); 
+            );
     }
 
-    isLoggedIn():boolean {
-    //return false if null or undefined, true otherwise!
+    isLoggedIn(): boolean {
+        //return false if null or undefined, true otherwise!
         console.log(
             "local user",
             localStorage.getItem('currentUser')
         );
-        return !!localStorage.getItem('currentUser');  
+        return !!localStorage.getItem('currentUser');
     }
-  
-    updateLocalStorageSubject():void {
+
+    updateLocalStorageSubject(): void {
         this.localStorageUserSubject.next(localStorage.getItem('currentUser'));
     }
-  
+
+    updateHabitCompletionLogs(habitCompletionLogs: IHabitCompletionLog[]) {
+        let completed: number = 0;
+        habitCompletionLogs.forEach(log => {
+
+            this.http.put<IHabitCompletionLog>(
+                this.habitCompletionLogsUrl,
+                log
+            ).
+                pipe(
+                    take(1),
+                    catchError(error => {
+                        return this.handleError(
+                            error,
+                            'Error: Failed to submit 1 or more habit logs'
+                        );
+                    })
+                ).
+                subscribe(() => {
+
+                    completed++;
+                    if (completed === habitCompletionLogs.length) {
+                        this.alertService.success('habit logs have been updated successfully');
+                    }
+                }
+
+                );
+        });
```

Comment:
This may work but there are better rxjs solutions for this. You are really creating a bunch of observables and want to act on the set but also track individuals. You should be able to report on the status of each individual observable as well as the set as a whole. To do this you can use `combineLatest`. You'll get updates for each observable and be able to handle failures of individuals. Here's an example if you'd like to tackle it.

``` TypeScript
combineLatest(selectedUsers.map(user => of({})
      .pipe(
        tap(() => user.progress = { loading: true, success: false, error: false }),
        switchMap(() => this.usersService.doSomething(user),
        tap(response => {
            const isThingDone = this.isThingDone(response);
            user.progress = { loading: false, success: isThingDone, error: !isThingDone }
        }),
        catchError(() => {
          user.progress = { loading: false, success: false, error: true };
          return of({ TotalThingsDone: 0 });
        })
      )
    )
))
```

Reply:
starting with `of({})` is typically used if you have a loading indicator that you want to show before you do anything else.